app/src/org/commcare/dalvik/activities/CommCareSetupActivity.java

```diff
@@ -600,17 +601,64 @@ private void scanSMSLinks(boolean installTriggeredManually){
         }
         int messageIterationCount = 0;
         try {
-            while (cursor.moveToNext() && messageIterationCount <= SMS_CHECK_COUNT) {
-                String textMessageBody = cursor.getString(cursor.getColumnIndex("body"));
+            while (cursor.moveToNext() && messageIterationCount <= SMS_CHECK_COUNT) { // must check the result to prevent exception
                 messageIterationCount++;
+                String textMessageBody = cursor.getString(cursor.getColumnIndex("body"));
                 if (textMessageBody.contains(GlobalConstants.SMS_INSTALL_KEY_STRING)) {
                     RetrieveParseVerifyMessageTask mTask =
-                            new RetrieveParseVerifyMessageTask(this, installTriggeredManually);
-                    mTask.execute(textMessageBody);
+                            new RetrieveParseVerifyMessageTask<CommCareSetupActivity>(this, installTriggeredManually) {
+
+                                @Override
+                                protected void deliverResult(CommCareSetupActivity receiver, String result) {
+                                    if (installTriggeredManually) {
+                                        if (result != null) {
+                                            incomingRef = result;

```

Comment:
All of this logic needs to be executed against the receiver, not the current class instance.

It's good to wrap these calls like

deliverResult() {
    receiver.doAction(result)
}

to make sure that happens reliably